### PR TITLE
Create Post Frontend and Backend

### PR DESCRIPTION
**Pull request by tbatk**
This PR implements the frontend and the backend for Journal Entries. On the frontend, users are able to give an image, title, description, category, and then write out a body paragraph using Quill. 
Screenshot_20201210_115604
When a user uploads a file, the backround of the file select changes to the image uploaded.
Screenshot_20201210_115550

When the publish button is clicked, a request to the API route `/api/journal/create` is sent. This route parses the form data, and uploads the data to Contentful. A new type, `JournalEntry`, has been added to make this process easier. 

I've also added in another backend function that allows us to retrieve entries by review status. This comes along with the API route `/api/journal/get`, which takes in a query from the url. By default, it will return the non-reviewed entries. You can also use `/api/journal/get?reviewed=false` to get non-reviewed entries. To get reviewed entries, you must make a request to `/api/journal/get?reviewed=true`. Please let me know if I need to make any changes.

### CODE REVIEW

File: components/CreateJournalEntry/index.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        }
+        
+    }
+    const handleQuill = (content, delta, source, editor) => {
+        //If the editor is empty, the only thing in it is a newline character. We don't want to send just newlines to the backend, so we do this.
+        if(editor.getText() != "\n"){
+            if(values.body){
+                delete values.error;
+            }
+            setValues(values => ({...values, ["body"]: editor.getHTML()}));
```
Comment (VishalAiely) [inline components/CreateJournalEntry/index.tsx:58]: Good error handling and validation

File: components/CreateJournalEntry/index.tsx [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+const ReactQuill = dynamic(import('react-quill'), { ssr: false});
+
+interface IFormValues {
+    title?: string,
+    description?: string,
+    image?: File,
+    body?: string,
+    category?: string, 
+    error?: boolean,
+    [key: string]: any, //TypeScript throws a weird error with this. This is set to any for now, but it can be changed later.
```
Comment (VishalAiely) [inline components/CreateJournalEntry/index.tsx:17]: I believe this is the correct way to handle this, try to make sure you cast it to a known type when used

File: server/actions/Contentful.ts [hunk 3]
```diff
@@ -45,13 +48,93 @@ export async function uploadImage(image: File){
...
+
+/**
+* @param journalEntry Journal Entry information to be uploaded to Contentful
+* @returns True if creation is successful, false if creation is unsuccessful.
+*/
+export async function createJournalEntry(journalEntry: JournalEntry){
+    try{
+        const space = await client.getSpace(process.env.CONTENTFUL_SPACE as string);
+        const environment = await space.getEnvironment('master');
+        const entry = await environment.createEntry('blogPost', {
```
Comment (VishalAiely) [inline server/actions/Contentful.ts:79]: We may want to store these strings in a utils file and import it so that we only have to change it once if needed

**Review by VishalAiely — APPROVED** [review]
I really like the function comments. Overall the code looks great to me; well written and easy to understand what is going on. We should run the linter through all files to have it autoformat all the documents but this will be done later to all files before production so don't worry about it. Thanks for all the hard work on the journal entries!

File: pages/api/journal/get.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { NextApiResponse, NextApiRequest } from "next";
+import {getJournalEntriesByReviewStatus} from "server/actions/Contentful";
+//This API route takes in a query from the url. By default, the route will return all entries that have not been reviewed. To get reviewed entries, use the following route: /api/journal/get?reviewed=true
+export default async function handler(req: NextApiRequest, res: NextApiResponse){
+
+    if(req.method == "GET"){
+        let reviewed = false;
+        if(req.query.reviewed){
+           reviewed = (req.query.reviewed === "true") || (req.query.reviewed === "false");
+        }
```
Comment (kemalfidan) [inline pages/api/journal/get.ts]: This bit is a little confusing. Don't we want to call getJournalEntriesByReviewStatus() with req.query.reviewed?

File: pages/api/journal/get.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { NextApiResponse, NextApiRequest } from "next";
```
Comment (kemalfidan) [inline pages/api/journal/get.ts]: I think this file should be renamed. This file gets blogs by review status and we'll have another one that gets blogs by a unique id.

**Review by kemalfidan — APPROVED** [review]

File: pages/api/journal/get.ts (hunk 4 above)
Comment (tbatk) [reply pages/api/journal/get.ts]: req.query.reviewed is a string, and the data type within Contentful is a boolean. So in order to retrieve posts by review status, req.query.reviewed has to be converted into a boolean.  I just realized that I messed up the assignment for reviewed, so I'll fix that really quick.

File: pages/api/journal/get.ts (hunk 5 above)
Comment (tbatk) [reply pages/api/journal/get.ts]: That's true, I didn't think about that. I'll fix that right way.

**Pull request merged**